cpp/dolfinx/common/array_2d.h

```diff
@@ -0,0 +1,187 @@
+// Copyright (C) 2021 Igor Baratta
+//
+// This file is part of DOLFINX (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#pragma once
+
+#include "span.hpp"
+#include <cassert>
+#include <vector>
+
+template <typename T, typename = void>
+struct is_2d_container : std::false_type
+{
+};
+
+template <typename T>
+struct is_2d_container<T, std::void_t<decltype(std::declval<T>().data()),
+                                      decltype(std::declval<T>().rows()),
+                                      decltype(std::declval<T>().cols())>>
+    : std::true_type
+{
+};
+
+namespace dolfinx::common
+{
+
+/// This class provides a dynamic 2-dimensional row-wise array
+/// data structure. The representation is strictly local, i.e.
+/// it is not parallel aware.
+template <typename T, class Allocator = std::allocator<T>>
+class array_2d
+{
+public:
+  /// \cond DO_NOT_DOCUMENT
+  using value_type = T;
+  using allocator_type = Allocator;
+  using size_type = typename std::vector<T, Allocator>::size_type;
+  using reference = typename std::vector<T, Allocator>::reference;
+  using const_reference = typename std::vector<T, Allocator>::const_reference;
+  using pointer = typename std::vector<T, Allocator>::pointer;
+  using iterator = typename std::vector<T, Allocator>::iterator;
+  using const_iterator = typename std::vector<T, Allocator>::const_iterator;
+  /// \endcond
+
+  /// Constructs an empty two dimensional array.
+  array_2d() : _rows(0), _cols(0) { _storage = std::vector<T, Allocator>(); }
```

Comment:
We can have `array_2d() = default;`, or explicitly create and empty array ..
It's used in a few places to return an empty array. 
eg: https://github.com/FEniCS/dolfinx/blob/master/cpp/dolfinx/mesh/cell_types.cpp#L139

Reply:
> We can have `array_2d() = default;`, or explicitly create and empty array ..
> It's used in a few places to return an empty array.
> eg: https://github.com/FEniCS/dolfinx/blob/master/cpp/dolfinx/mesh/cell_types.cpp#L139

It would be handy to be explicit with the row/column size - sometimes even if the array is empty knowing the number of columns (or rows) carries useful information.